### ISSUE TITLE
Support nonOptionalTypename: false

### ISSUE THREAD

**Issue by ertrzyiks** [zhouzi/graphql-codegen-factories#49]
By default the __typename field is added to the queries as optional. It is not compatible with extraction code that is currently used for example `Extract<BlogListQueryQuery, { __typename: 'Query' }>`

I had to enable my codegen config:
```
config:
  nonOptionalTypename: true
```

I suggest the plugin either to recommend using `nonOptionalTypename: true` or support typename field being optional.

**Comment by zhouzi** [comment]
Good catch, thanks.

Ideally I'd like not to have the code be based on the `__typename` field as I know some developers completely disable it through other plugins' option. But then I don't know how to extract a type from an union.

With #45 the use of `Extract` will be limited to unions instead of being used as soon as there's a type condition. And perhaps we can add a note in the documentation that `nonOptionalTypename` must be set to `true` if the developer is using unions. We could throw an error as well if the plugin tries to generate a union factory when this option is not set to `true`.

What do you think? Would that be acceptable?

**Comment by zhouzi** [comment]
I've been thinking about this issue and I may have found a solution that doesn't require developers to change their configuration:

```diff
- Extract<BlogListQueryQuery, { __typename: 'Query' }>
+ Extract<BlogListQueryQuery & { __typename: NonNullable<BlogListQueryQuery["__typename"]> }, { __typename: 'Query' }>
```

**Comment by zhouzi** [comment]
This has been fixed by #45 and will be published with a `1.0.0`.
